stream/src/main/java/org/apache/kafka/streaming/processor/internals/ProcessorContextImpl.java

```diff
@@ -0,0 +1,259 @@
+
+import org.apache.kafka.clients.consumer.Consumer;
+import org.apache.kafka.clients.consumer.KafkaConsumer;
+import org.apache.kafka.common.KafkaException;
+import org.apache.kafka.common.TopicPartition;
+import org.apache.kafka.common.metrics.Metrics;
+import org.apache.kafka.common.serialization.ByteArrayDeserializer;
+import org.apache.kafka.common.serialization.Deserializer;
+import org.apache.kafka.common.serialization.Serializer;
+import org.apache.kafka.streaming.StreamingConfig;
+import org.apache.kafka.streaming.processor.ProcessorContext;
+import org.apache.kafka.streaming.processor.StateStore;
+import org.apache.kafka.streaming.processor.Processor;
+import org.apache.kafka.streaming.processor.RestoreFunc;
+
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.IOException;
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+public class ProcessorContextImpl implements ProcessorContext {
+
+    private static final Logger log = LoggerFactory.getLogger(ProcessorContextImpl.class);
+
+    private final int id;
+    private final StreamTask task;
+    private final Metrics metrics;
+    private final RecordCollector collector;
+    private final ProcessorStateManager stateMgr;
+    private final ArrayDeque<ProcessorNode> nodeStack = new ArrayDeque<ProcessorNode>();
+
+    private final Serializer<?> keySerializer;
+    private final Serializer<?> valSerializer;
+    private final Deserializer<?> keyDeserializer;
+    private final Deserializer<?> valDeserializer;
+
+    private boolean initialized;
+
+    @SuppressWarnings("unchecked")
+    public ProcessorContextImpl(int id,
+                                StreamTask task,
+                                StreamingConfig config,
+                                RecordCollector collector,
+                                Metrics metrics) throws IOException {
+        this.id = id;
+        this.task = task;
+        this.metrics = metrics;
+        this.collector = collector;
+
+        this.keySerializer = config.getConfiguredInstance(StreamingConfig.KEY_SERIALIZER_CLASS_CONFIG, Serializer.class);
+        this.valSerializer = config.getConfiguredInstance(StreamingConfig.VALUE_SERIALIZER_CLASS_CONFIG, Serializer.class);
+        this.keyDeserializer = config.getConfiguredInstance(StreamingConfig.KEY_DESERIALIZER_CLASS_CONFIG, Deserializer.class);
+        this.valDeserializer = config.getConfiguredInstance(StreamingConfig.VALUE_DESERIALIZER_CLASS_CONFIG, Deserializer.class);
+
+        File stateFile = new File(config.getString(StreamingConfig.STATE_DIR_CONFIG), Integer.toString(id));
+
+        Consumer restoreConsumer = new KafkaConsumer<>(
+            config.getConsumerConfigs(),
+            null /* no callback for restore consumer */,
+            new ByteArrayDeserializer(),
+            new ByteArrayDeserializer());
```

Comment:
The fact that this class is creating its own Consumer is a problem for test cases, especially those that directly use `StreamTask`. I'd love to introduce a `ConsumerSupplier` (in Java 8 it'd simply be `Supplier<Consumer>`, but alas) and pass an implementation into this constructor and actually into the `StreamTask` constructor. This would then move the creation of this `ConsumerSupplier` into the `StreamThread`, which is already creating the `KafkaProducer` and `KafkaConsumer` instances used to consume records to pass to the `StreamTask`. And if this is acceptable, is it still desirable to use a *separate* `KafkaConsumer` instance for the `ProcessorStateManager`?